### ISSUE TITLE
Port remaining code in `twisted.python.deprecate` to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6036]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6036|
|Type|enhancement|
|Created|2012-10-01 15:00:13Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-py3-6036-2|

Parts of trial depend on the remaining code in `twisted.python.deprecate` that hasn't yet been ported to Python 3; all of it should be ported, or that dependency removed.

<details><summary>Searchable metadata</summary>

```
trac-id__6036 6036
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_deprecate_py3_6036_2 branches/deprecate-py3-6036-2
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1349103613000000 1349103613000000
changetime__1349294715000000 1349294715000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35888]) Branching to 'deprecate-py3-6036'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-py3-6036

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

Should be split into separate tickets for `twisted.python.modules` changes and `twisted.python.deprecate` changes, as discussed on IRC.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [35903]) Branching to 'deprecate-py3-6036-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun|
|-|-|

Merged forward with modules change from [#6041](https://github.com/twisted/twisted/issues/6041).

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-py3-6036-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. `twisted/python/test/deprecatedattributes.py` could use a copyright header and a module docstring
  1. Also in deprecatedattributes.py, the comment above the reflect import is obsolete - there is no longer a circular dependency between those two modules (even disregarding Python 3 porting artifacts).
  1. *Also* reflect isn't actually used once it's imported.
  1. Another unused import, `traceback`, in `twisted/python/reflect.py` - not new in this branch though.
  1. in `test_deprecate.py`, `ImportedModuleAttributeTests.pathEntryTree` first encodes a path string using UTF-8 and then potentially tries to decode it using ASCII by way of `nativeString`.  This is fine until we encounter a non-ASCII path.  Not that I really care about non-ASCII paths, but I think we might have decided to support them (even before `FilePath` supports them).  Not supporting them would be easier, though. :)  This came up in `twisted.trial.test.test_warnings` and apparently the solution was to just explicitly decode using UTF-8 and put the result into `sys.path` (ie, not try to use `nativeString` for this case).  That's in trunk now, but I have no idea if it actually works.
  1. Oops, ditto that problem for `checkOneWarning`.
  1. And `WarnAboutFunctionTests.setUp`.
  1. For deciding whether to call `invalidate_caches` or not, perhaps a try/except ImportError is better than a `_PY3` check.  Again, this is what the code in `twisted/trial/test/test_warnings.py` does.

Nothing major here except perhaps the `nativeString` issue, the resolution to which I'm curious about, so I'd like to look at that, at least, before this is merged.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

(In [35921]) Address review comments. Refs [#6036](https://github.com/twisted/twisted/issues/6036)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

(In [35922]) Address review comments, in the correct branch. Refs [#6036](https://github.com/twisted/twisted/issues/6036)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set owner to @exarkun|
|-|-|

OK, fixed, in a number of ways depending on `nativeString` case.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-py3-6036-2 is running.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @itamarst|
|-|-|

Okay, great, as long as buildbot is happy (the documentation builder isn't, but I don't think that's actually a new error).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set status to `closed`|
|-|-|

(In [35927]) Merge deprecate-py3-6036-2.

Author: itamar
Review: exarkun
Fixes: [#6036](https://github.com/twisted/twisted/issues/6036)

twisted.python.deprecate now runs on Python 3.
